# **IRONHACK_Sharks**

In [ ]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')  
pd.set_option('display.max_columns', None)  
from src.shark_func import *

Proyecto de limpieza del archivo attacks.csv, que se puede localizar en el siguiente link https://www.kaggle.com/datasets/teajay/global-shark-attacks/download?datasetVersionNumber=7

In [48]:
# attacks.csv no está codificado en UTF-8, hay que especificar la codificación en la importación
sharks = pd.read_csv('../data/attacks.csv',encoding='ANSI')    

In [49]:
shape_origin = sharks.shape         # Guardamos el tamaño inicial del dataset
sharks.head(10)

Case Number         Date    Year        Type    Country             Area   
0    2018.06.25  25-Jun-2018  2018.0     Boating        USA       California  \
1    2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2    2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3    2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4    2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   
5  2018.06.03.b  03-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
6  2018.06.03.a  03-Jun-2018  2018.0  Unprovoked     BRAZIL       Pernambuco   
7    2018.05.27  27-May-2018  2018.0  Unprovoked        USA          Florida   
8  2018.05.26.b  26-May-2018  2018.0  Unprovoked        USA          Florida   
9  2018.05.26.a  26-May-2018  2018.0  Unprovoked        USA          Florida   

                                            Location      Activity   
0                        Oceanside, San Diego County      Paddling  \
1                     St. Simon Island, Glynn County      Standing   
2                                       Habush, Oahu       Surfing   
3                                 Arrawarra Headland       Surfing   
4                                           La Ticla   Free diving   
5                                 Flat Rock, Ballina  Kite surfing   
6                              Piedade Beach, Recife      Swimming   
7  Lighhouse Point Park, Ponce Inlet, Volusia County       Fishing   
8                       Cocoa Beach, Brevard  County       Walking   
9                      Daytona Beach, Volusia County      Standing   

                     Name Sex   Age   
0             Julie Wolfe    F   57  \
1         Adyson McNeely     F   11   
2             John Denges    M   48   
3                    male    M  NaN   
4          Gustavo Ramos     M  NaN   
5                 Chris …    M  NaN   
6  Jose Ernesto da Silva     M   18   
7                    male    M   52   
8               Cody High    M   15   
9                    male    M   12   

                                              Injury Fatal (Y/N)   
0  No injury to occupant, outrigger canoe and pad...           N  \
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   
5                            No injury, board bitten           N   
6                                              FATAL           Y   
7            Minor injury to foot. PROVOKED INCIDENT           N   
8                              Lower left leg bitten           N   
9                               Minor injury to foot           N   

             Time         Species             Investigator or Source   
0           18h00      White shark                  R. Collier, GSAF  \
1   14h00  -15h00              NaN    K.McMurray, TrackingSharks.com   
2           07h45              NaN    K.McMurray, TrackingSharks.com   
3             NaN        2 m shark                    B. Myatt, GSAF   
4             NaN  Tiger shark, 3m                         A .Kipper   
5             NaN              NaN         Daily Telegraph, 6/4/2018   
6  Late afternoon      Tiger shark    Diario de Pernambuco, 6/4/2018   
7             NaN  Lemon shark, 3'   K. McMurray, TrackingSharks.com   
8           17h00   Bull shark, 6'    K.McMurray, TrackingSharks.com   
9           14h00              NaN  K. McMurray, Tracking Sharks.com   

                             pdf   
0           2018.06.25-Wolfe.pdf  \
1         2018.06.18-McNeely.pdf   
2          2018.06.09-Denges.pdf   
3       2018.06.08-Arrawarra.pdf   
4           2018.06.04-Ramos.pdf   
5      2018.06.03.b-FlatRock.pdf   
6       2018.06.03.a-daSilva.pdf   
7           2018.05.27-Ponce.pdf   
8          2018.05.26.b-High.pdf   
9  2018.05.26.a-Dayt

In [50]:
sharks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

#### PRIMERAS IMPRESIONES:
* **Case Number:** identificación única del caso (tip formatearlo de alguna manera para poder usar la fecha)
* **Date:** Fecha
* Year
* Type  Tiene pinta que es el ataque (value counts provoke unprovoke unknown)
* Country
* Area
* Location
* Name
* Sex
* Age
* Injury
* Fatal (yes or no/ Unknown)
* Time
* Species (aquí hay chica)
* Investigator or Source: posibilidad de pasarla a dos columnas
* pdf
* href formula
* href duplicada
* Case number 1
* Case number 2
* original order: WTF
* Unnamed 22: Las voy a usar para otras cosas
* Unnamed 23: Las voy a usar para otras cosas

Vamos a comprobar si tenemos filas duplicadas

In [51]:
sharks.duplicated().sum()

19411

Tenemos 19411 lineas duplicadas. Procedemos a eliminarlas. No aportan información alguna

In [52]:
sharks.drop_duplicates(inplace = True)

In [53]:
sharks.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Index: 6312 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             6310 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null   obje

In [55]:
check_nan_cols(sharks);

Case Number                  2
Date                        10
Year                        12
Type                        14
Country                     60
Area                       465
Location                   550
Activity                   554
Name                       220
Sex                        575
Age                       2841
Injury                      38
Fatal (Y/N)                549
Time                      3364
Species                   2848
Investigator or Source      27
pdf                         10
href formula                11
href                        10
Case Number.1               10
Case Number.2               10
original order               3
Unnamed: 22               6311
Unnamed: 23               6310
dtype: int64

Emepezamos por cols 22 y 23. Son todo nulos salvo 3 valores

In [27]:
sharks[(sharks['Unnamed: 22'].notna()) | sharks['Unnamed: 23'].notna() ]

Case Number                  Date    Year        Type   
1478    2006.05.27           27-May-2006  2006.0  Unprovoked  \
4415    1952.03.30           30-Mar-1952  1952.0  Unprovoked   
5840  1878.09.14.R  Reported 14-Sep-1878  1878.0    Provoked   

                   Country         Area                    Location   
1478                   USA       Hawaii          North Shore, O'ahu  \
4415  NETHERLANDS ANTILLES      Curacao                         NaN   
5840                   USA  Connecticut  Branford, New Haven County   

                                             Activity              Name Sex    
1478                                          Surfing      Bret Desmond    M  \
4415  Went to aid of child being menaced by the shark       A.J. Eggink    M   
5840                                          Fishing  Captain Pattison    M   

      Age                                        Injury Fatal (Y/N)   Time   
1478   31             No injury, shark bumped surfboard           N  16h00  \
4415  NaN                Buttock bitten, tissue removed           N    NaN   
5840  NaN  Leg bitten by netted shark PROVOKED INCIDENT           N    NaN   

                                               Species    
1478                                                NaN  \
4415  Bull shark, 2.7 m [9'] was captured & dragged ...   
5840                                                NaN   

                                 Investigator or Source   
1478                                         R. Collier  \
4415  J. Randall, p.352 in Sharks & Survival; H.D. B...   
5840                       St. Joseph Herald, 9/14/1878   

                            pdf   
1478     2006.05.27-Desmond.pdf  \
4415      1952.03.30-Eggink.pdf   
5840  1878.09.14.R-Pattison.pdf   

                                           href formula   
1478  http://sharkattackfile.net/spreadsheets/pdf_di...  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1   
1478  http://sharkattackfile.net/spreadsheets/pdf_di...    2006.05.27  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...    1952.03.30   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...  1878.09.14.R   

     Case Number.2  original order   Unnamed: 22      Unnamed: 23  
1478    2006.05.27          4825.0  stopped here              NaN  
4415    1952.03.30          1888.0           NaN           Teramo  
5840  1878.09.14.R           463.0           NaN  change filename

19411